### ISSUE TITLE
Mathematical inequivalence introduced by GLT Sampler vs. DGL Sampler? 

### ISSUE THREAD

**Issue by Elnifio** [alibaba/graphlearn-for-pytorch#79]
### 🐛 Describe the bug

We observe that there are some differences in samplers that leads to computational / mathematical inequivalences between GLT's implementation of IGBH training vs. IGB's "official" implementation, when running the training code. More specifically, we observe that, compared to IGB's DGL sampler, GLT/PyG sampler leads to unnecessary message passing & aggregation and therefore results in different computation results. 

In a high level, there are two differences that leads to the computational inequivalence here: 
1. Given a path `a->b->c`, at each layer, GLT is updating all nodes at this layer and higher layer. For instance, at layer 0, we need to update node embeddings for both b and c. However, in DGL, only node b's embedding will be updated. Therefore, if we consider the final embedding generated for node c, GLT computes it as `h_2(c) = h_1(c) + h_1(b) = (h_0(c) + h_0(b)) + (h_0(b) + h_0(a))`, while DGL computes it as `h_2(c) = h_1(c) + h_1(b) = h_0(c) + (h_0(b) + h_0(a))`.
2. When creating blocks for layer $i-1$, DGL not only uses the destination nodes from layer $i$ as seed nodes, but also includes all subsequent $i+1\cdots n$ destination nodes as seed nodes.

Take the following graph as an example: 

```
3 -> 9, 9 -> 12, 8 -> 12
```

If the path `3 -> 9 -> 12` is being sampled, then the message passing path in GLT is: 
1. at first layer, we compute `3 -> 9` and `9 -> 12`
2. at second layer, we compute `9 (value updated with node 3's feature) -> 12`

Thus, from the above computation, we see that feature 9 is included twice.

However, in DGL, the sampling process is a bit different: 
1. We start with a seed node set containing only our target node 12: `[12, ] `
2. At first hop (corresponding to the model's second layer), we add 9, since 9 is randomly chosen as 12's source node, added to the seed node set: `[12, 9]`
3. At second hop (corresponding to the model's first layer), we add:
  1. node 8, since node 12 is in the seed node set
  2. node 3, since node 9 is in the seed node set
4. And therefore we results in `[12, 9, 8, 3]` nodes, with graph `3 -> 9 -> 12 <- 8`. 

Therefore, in the first layer, we compute `3 -> 9` and `8 -> 12`; and in the second layer, we additionally compute `9 -> 12`. This computation is the algorithm that is implemented in the paper (IGB's official repository), and is mathematically different from GLT's current implementation. 

### Environment

* GLT version: latest, built from main branch. 
* PyG version: 2.3.1
* PyTorch version: 1.14.0
* OS: Linux
* Python version: 3.8.10
* CUDA/cuDNN version: 11.8
* Any other relevant information

**Comment by Elnifio** [comment]
To reproduce the issue, we run the following code snippet in GLT:

```
import graphlearn_torch as glt
import torch
import torch_geometric as pyg

# We consider a simplified version of GATConv here. 
# GATConv computes the destination nodes' embedding from both its embedding and its source nodes' embedding
# we simplify it here so that only an addition operation is performed
# between destination nodes' embedding and the source nodes' embedding. 
class DummyLayer(pyg.nn.MessagePassing):
    def __init__(self):
        super().__init__(aggr="add")
        
    def forward(self, x, edge_index):
        return self.propagate(edge_index, x=x)
    
    def message(self, x_i, x_j):
        return x_i + x_j

# to simplify the discussion, we create a tree for generating the samples. 
u = torch.tensor([0, 1, 2, 3,  4,  5,  6,  7,  8,  9, 10, 11])
v = torch.tensor([8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13])
edge_index = torch.stack([u, v], dim=0)

# and to trace which node is involved in the computation,
# the features are assigned as 2^i for node i
x = torch.stack([2**torch.tensor(range(14)) for _ in range(4)], dim=1).to(torch.float32)

# GLT IGBH's dataloading code
dataset = glt.data.Dataset(edge_dir="in")
dataset.init_graph(edge_index, graph_mode="CPU")
dataset.init_node_features(node_feature_data=x, with_gpu=False)
loader = glt.loader.NeighborLoader(
    dataset,
    [1, 1],
    torch.tensor([12, 13]),
    batch_size=2
)
data = next(iter(loader))

x = data.x
edge_index = data.edge_index
# At start, the edge index could be: 
# 2 -> 9 -> 12
# 5 -> 10 -> 13
# and the corresponding features could be:
# tensor([[4096, 4096, 4096, 4096], # node 12's feature
#         [8192, 8192, 8192, 8192], # node 13's feature
#         [ 512,  512,  512,  512], # node 9's feature
#         [1024, 1024, 1024, 1024], # node 10's feature
#         [   4,    4,    4,    4], # node 2's feature
#         [  32,   32,   32,   32]], dtype=torch.int32) # node 5's feature

h1 = DummyLayer()(x, edge_index)
# after we apply the first transformation, the features now are updated as:
# tensor([[4608, 4608, 4608, 4608], # node 12, h_1(12) = h_0(12) + h_0(9)
#         [9216, 9216, 9216, 9216], # node 13, h_1(13) = h_0(13) + h_0(10)
#         [ 516,  516,  516,  516], # node 9, h_1(9) = h_0(9) + h_0(2)
#         [1056, 1056, 1056, 1056], # node 10, h_1(10) = h_0(10) + h_0(5)
#         [   0,    0,    0,    0], # node 2, no update since there are no incoming edges
#         [   0,    0,    0,    0]], dtype=torch.int32) # node 5, no update since there are no incoming edges

h2 = DummyLayer()(h1, edge_index)
# after we apply the second transformation, the features now are updated as:
# tensor([[ 5124,  5124,  5124,  5124], # node 12, h_2(12) = h_1(12) + h_1(9)
#         [10272, 10272, 10272, 10272], # node 13, h_2(13) = h_1(13) + h_1(10)
#         [  516,   516,   516,   516],
#         [ 1056,  1056,  1056,  1056],
#         [    0,     0,     0,     0],
#         [    0,     0,     0,     0]], dtype=torch.int32)

# for node 12, the message passing path is:
# 9 -> 12
# 2 -> 9 (value updated with node 2's feature) -> 12
# thus, 9 is updated twice. 

# this is different from DGL's implementation
# in DGL, if we have a path 2 -> 9 -> 12, 
# then the value will be: (2**12 + 2**9 + 2**2)
# verified:
# (2**12 + 2**9 + (2**9 + 2**2)) == 5124
```

And we run the following code for DGL:
```
import dgl
import torch

class DummyLayer(torch.nn.Module):
    def __init__(self):
        # Takes in source and destination nodes, adds them according to the edges, and returns it
        super(DummyLayer, self).__init__()
        
    def forward(self, graph, hidden):
        graph = graph.local_var()
        graph.srcdata['temp_src'] = hidden
        graph.dstdata['temp_dst'] = hidden[:graph.number_of_dst_nodes()]
        graph.update_all(
            dgl.function.u_add_v('temp_src', 'temp_dst', 'temp'),
            dgl.function.sum("temp", "result")
        )
        new_hidden = graph.dstdata['result']
        return new_hidden

u = torch.tensor([0, 1, 2, 3,  4,  5,  6,  7,  8,  9, 10, 11])
v = torch.tensor([8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13])
g = dgl.graph((u, v))

g.ndata['feat'] = torch.stack(
    [2**g.nodes(), 2**g.nodes(), 2**g.nodes(), 2**g.nodes()], 
    dim=1).to(torch.float32) 

sampler = dgl.dataloading.NeighborSampler([1, 1])
loader = dgl.dataloading.DataLoader(g, torch.tensor([12, 13]), sampler, batch_size=2)
input_nodes, destination_nodes, blocks = next(iter(loader))

x = blocks[0].srcdata['feat']
# here, if we examine the edges in block 0, we might see: 
# 8->12
# 10->13
# 2->9
# 4->10
# only these destination nodes' information are being updated. 
# and we examine the features: 
# tensor([[4096, 4096, 4096, 4096], # node 12
#         [8192, 8192, 8192, 8192], # node 13
#         [ 512,  512,  512,  512], # node 9
#         [1024, 1024, 1024, 1024], # node 10
#         [ 256,  256,  256,  256], # node 8
#         [   4,    4,    4,    4], # node 2
#         [  16,   16,   16,   16]], dtype=torch.int32) # node 4

h1 = DummyLayer()(blocks[0], x)
# After applying the transformation along edges defined in block 0, we see the information is updated as:
# tensor([[4352, 4352, 4352, 4352], # node 12, as a sum of node 8 and node 12
#         [9216, 9216, 9216, 9216], # node 13, as a sum of node 10 and node 13
#         [ 516,  516,  516,  516], # node 9, as a sum of node 2 and node 9
#         [1040, 1040, 1040, 1040]], dtype=torch.int32) # node 10, as a sum of node 4 and node 10

# Now, if we examine block 1, we should see edges: 
# 9->12
# 10->13

h2 = DummyLayer()(blocks[1], h1)
# After applying the transformation along edges defined in block 1, we see the information is updated as:
# tensor([[ 4868,  4868,  4868,  4868], # final result for node 12
#         [10256, 10256, 10256, 10256]], dtype=torch.int32) # final result for node 13

# for node 12, the sampling is: 
# we start with [12]
# at first hop, we add 9: [12, 9]
# at second hop, we add: 
#     8, since 12 is in the collection of seed node (this feature is discussed in DGL thread)
#     2, since 9 is in the collection of seed node
# so we have [12, 9, 2, 8]
# and the message passing path that DGL takes is: 
# 2 -> 9 -> 12 <- 8
# so, the information that we get, should be all information that nodes 2, 8, 9, and 12 have:

# verified
# (2**9 + 2**12 + 2**8 + 2**2) == 4868
```

**Comment by LiSu** [comment]
Hi @Elnifio，

Yes, this difference is caused by the different approaches of sampling and message passing in PyG and DGL.

**Comment by hesham-mostafa** [comment]
I believe the two sampling approaches (GLT's and DGL's) would lead to mathematically equivalent outputs at the seed nodes in the top layer. See this thread on the PyG discussions: [https://github.com/pyg-team/pytorch_geometric/discussions/5303](https://github.com/pyg-team/pytorch_geometric/discussions/5303)  . 

GLT's ``NeighborLoader`` introduces some redundant nodes in each layer that are not needed for producing the output at the seed nodes. The GNNs applied to GLT's sampled graphs would thus require more FLOPs/memory accesses than the GNNs applied to DGL's sampled blocks.  I modified the code above to show the equivalence in the produced outputs:

```python

import graphlearn_torch as glt
import torch
import torch_geometric as pyg
import dgl

# to simplify the discussion, we create a tree for generating the samples. 
u = torch.tensor([0, 1, 2, 3,  4,  5,  6,  7,  8,  9, 10, 11])
v = torch.tensor([8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13])
edge_index = torch.stack([u, v], dim=0)

# and to trace which node is involved in the computation,
# the features are assigned as 2^i for node i
x = torch.stack([2**torch.tensor(range(14)) for _ in range(4)], dim=1).to(torch.float32)


class DummyLayerDGL(torch.nn.Module):
    def __init__(self):
        # Takes in source and destination nodes, adds them according to the edges, and returns it
        super().__init__()
        
    def forward(self, graph, hidden):
        graph = graph.local_var()
        graph.srcdata['temp_src'] = hidden[:graph.number_of_src_nodes()]
        graph.dstdata['temp_dst'] = hidden[:graph.number_of_dst_nodes()]
        graph.update_all(
            dgl.function.u_add_v('temp_src', 'temp_dst', 'temp'),
            dgl.function.sum("temp", "result")
        )
        new_hidden = graph.dstdata['result']
        return new_hidden

class DummyLayerGLT(pyg.nn.MessagePassing):
    def __init__(self):
        super().__init__(aggr="add")
        
    def forward(self, x, edge_index):
        return self.propagate(edge_index, x=x)
    
    def message(self, x_i, x_j):
        return x_i + x_j


# GLT 
dataset = glt.data.Dataset(edge_dir="in")
dataset.init_graph(edge_index, graph_mode="CPU")
dataset.init_node_features(node_feature_data=x, with_gpu=False)
loader = glt.loader.NeighborLoader(
    dataset,
    [-1, -1],
    torch.tensor([12, 13]),
    batch_size=1
)
data = next(iter(loader))

h1_glt = DummyLayerGLT()(x, edge_index)
h2_glt = DummyLayerGLT()(h1_glt, edge_index)


#DGL
g = dgl.graph((u,v))
g.ndata['feat'] = x
sampler = dgl.dataloading.NeighborSampler([-1, -1])
loader = dgl.dataloading.DataLoader(g, torch.tensor([12, 13]), sampler, batch_size=2)
input_nodes, destination_nodes, blocks = next(iter(loader))

h1_dgl = DummyLayerDGL()(blocks[0],x[input_nodes])
h2_dgl = DummyLayerDGL()(blocks[1],h1_dgl)

#equivalence
assert torch.allclose(h2_dgl,h2_glt[[12,13]])

```
